### ISSUE TITLE
Konnector for ooshop

### ISSUE THREAD

**Issue by eravion** [cozy-labs/konnectors#81]
Get invoice pdf from ooshop : www.ooshop.com

**Comment by eravion** [comment]
I asked to weboob team, with our answer
https://weboob.org/issues/1873

**Comment by ZeHiro** [comment]
Writing a konnector is not hard, if you know a bit of coding, I can help you develop one if necessary.

**Comment by eravion** [comment]
I know a bit of coding, but not in Java ...
I'm mainly scripting in bash for sysadmin tasks.
I'll check ooshop web site to understand how to get pdf invoiced.
If I can develop something in bash, I'll explain you how it works and, maybe we can develop a Konnector together.

**Comment by frankrousseau** [comment]
NB: We can provide assitance for coding a konnector. :)

**Comment by eravion** [comment]
I'll write whet I have.
One remaining point is to hack the password ...

First login to get the cli_id
Need : login (email of customer) and password
Pb password is scrambled - TO reverse encryption system pwd toto = 65E1A5E1C0000000
http://mshop.carrefour.com/convertigo/projects/ooshop/.json?__sequence=CheckLogin&login=${LOGIN}&password={PASSWORD}
From the answer, get the cli_id 

Get list of all invoices 
http://mshop.carrefour.com/convertigo/projects/ooshop/.json?__sequence=GetMyAccount&cli_id=${cli_id} 
get the id and for each

Get product list for an invoice
http://mshop.carrefour.com/convertigo/projects/ooshop/.json?__sequence=GetCmdProduct&cli_id=${cli_id}&id=${id}
Output is text, not pdf
One interesting information is in the gif name
the name is the EAN code which can be used with http://fr.openfoodfacts.org/
sample :   "img_product_url": "Media\/ProdImages\/Produit\/Images\/3181232220163.gif",

**Comment by ZeHiro** [comment]
Can't we get the invoices through the regular ooshop website? We would get the pdf this way no?
HAcking the connection process of the wesite wouldn't be to hard. Then  the konnector just go through the webpage and look for the invoices, before downloading them.

**Comment by eravion** [comment]
Yes it looks possible to go through the web site, but it is aspx and I don't understand how to connect.
If you want to test, you can create an account for free.
Second point is that the web site is not providing pdf. The invoice are available in texte mode by a function used to print the invoice. once again this is aspx.

The site used by Apps is more easy to understand for me. Despite the password, it is ok.
For information invoice are not protected, if you have the cli_id and the invoice number you can get the invoice without cookie or authentification.
If you want to try to connect to the web site, the information needed is this cli_id 
        <input type="hidden" id="hidOmnitureUtilisateurID" value="${cli_id}" />\r\n

I'm on IRC if you want to exchange by chat, do not hesitate.

**Comment by ZeHiro** [comment]
Info to import from basket : 
 "id": "96568",
"title": "Framboise Cat 1",
 "uvc_buy": "3.50",
 "contenance": "la barquette de 125g",
"uvc_info": "26.92 € \/ kg",
"img_product_url": "Media\/ProdImages\/Produit\/Images\/3660239015488.gif", => TO CHANGE TO EAN
 "is_dispo": "true",
"art_qte": "4"

**Comment by eravion** [comment]
Hello, 

More information about "howto" login

1/ get the content of __VIEWSTATE variable and the content of __EVENTVALIDATION variable
curl "http://www.ooshop.com/courses-en-ligne/WebForms/Utilisateur/VerifEligibilite.aspx"  | egrep "(__VIEWSTATE|__EVENTVALIDATION)"

2/ in both variable replace all special character by web character
/ by %2F
+ by %2B

3/ login and get cli_id
We have 4 variables
USER  (come from the UI)
PASSWORD (come from the UI)
VIEWSTATE (see point 1 and 2)
EVENTVALIDATION (see point 1 and 2)
curl "http://www.ooshop.com/courses-en-ligne/WebForms/Utilisateur/VerifEligibilite.aspx" \
-d "ctl00%24sm=ctl00%24sm%7Cctl00%24cphC%24elig%24lv%24LoginButton&__EVENTTARGET=&__EVENTARGUMENT=&__LASTFOCUS=&__VIEWSTATE=${VIESTATE}&__EVENTVALIDATION=${EVENTVALIDATION}&ctl00%24xCoordHolder=0&ctl00%24yCoordHolder=0&ctl00%24headerCtrl%24txtB=&ctl00%24cphC%24elig%24lv%24utilisateur=${USER}&ctl00%24cphC%24elig%24lv%24mot_pass=${PASSWORD}&ctl00%24cphC%24elig%24txtFinishCmd=False&ctl00%24cphC%24elig%24tbCp=&ctl00%24cphC%24elig%24login2=&ctl00%24ucDetProd%24AjoutPanier1%24nbrProduit=1&ctl00%24Perso%24ucAu%24lv%24UserName=Email%20ou%20Login&ctl00%24Perso%24ucAu%24lv%24Password=&ctl00%24Perso%24ucAu%24login=&ctl00%24Perso%24ucAu%24tbEmail=&ctl00%24cphC%24elig%24lv%24LoginButton=VALIDER" | grep s.eVar5

s.eVar5 = cli_id for the next request

**Comment by frankrousseau** [comment]
@ZeHiro I wrote a first draft of a documentation about how to write konnectors. It could help you: https://github.com/cozy-labs/konnectors/wiki

**Comment by NTag** [comment]
I have no idea where you are in this issue, but you can use the "JSON API", and give the password in plain-text to log in (with this URL: http://mshop.carrefour.com/convertigo/projects/ooshop/.json?__sequence=CheckLogin&login=${LOGIN}&password={PASSWORD} )

No need to find the hash method they are using.

**Comment by frankrousseau** [comment]
By the way you will soon be able to write your konnectors with Javascript. Coffeescript won't be mandatory anymore to write connectors.

**Comment by teolemon** [comment]
Really cool. Feel free to join us on Open Food Facts's Slack to see how we can help.
